polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1256,31 +1330,359 @@ impl<T: Config> Pallet<T> {
 		let value = (origin_location, assets.into_inner());
 		ensure!(T::XcmTeleportFilter::contains(&value), Error::<T>::Filtered);
 		let (origin_location, assets) = value;
+		for asset in assets.iter() {
+			let transfer_type =
+				<T::XcmExecutor as AssetTransferSupport>::determine_for(asset, &dest)
+					.map_err(Error::<T>::from)?;
+			ensure!(matches!(transfer_type, TransferType::Teleport), Error::<T>::Filtered);
+		}
+		let fees = assets.get(fee_asset_item as usize).ok_or(Error::<T>::Empty)?.clone();
+
+		Self::build_and_execute_xcm_transfer_type(
+			origin_location,
+			dest,
+			beneficiary,
+			assets,
+			TransferType::Teleport,
+			fees,
+			None,
+			weight_limit,
+		)
+	}
+
+	fn build_and_execute_xcm_transfer_type(
+		origin: MultiLocation,
+		dest: MultiLocation,
+		beneficiary: MultiLocation,
+		assets: Vec<MultiAsset>,
+		transfer_type: TransferType,
+		fees: MultiAsset,
+		prefund_fees_messages: Option<(Xcm<<T as Config>::RuntimeCall>, Xcm<()>)>,
+		weight_limit: WeightLimit,
+	) -> DispatchResult {
+		let (mut local_xcm, remote_xcm) = match transfer_type {
+			TransferType::LocalReserve => {
+				let (local, remote) = Self::local_reserve_transfer_messages(
+					dest,
+					beneficiary,
+					assets,
+					fees,
+					prefund_fees_messages,
+					weight_limit,
+				)?;
+				(local, Some(remote))
+			},
+			TransferType::DestinationReserve => {
+				let (local, remote) = Self::destination_reserve_transfer_messages(
+					dest,
+					beneficiary,
+					assets,
+					fees,
+					prefund_fees_messages,
+					weight_limit,
+				)?;
+				(local, Some(remote))
+			},
+			TransferType::RemoteReserve(reserve) => (
+				Self::remote_reserve_transfer_message(
+					reserve,
+					dest,
+					beneficiary,
+					assets,
+					fees,
+					weight_limit,
+				)?,
+				None,
+			),
+			TransferType::Teleport =>
+				(Self::teleport_asset_message(dest, beneficiary, assets, fees, weight_limit)?, None),
+		};
+		let weight =
+			T::Weigher::weight(&mut local_xcm).map_err(|()| Error::<T>::UnweighableMessage)?;
+		let hash = local_xcm.using_encoded(sp_io::hashing::blake2_256);
+		let outcome =
+			T::XcmExecutor::execute_xcm_in_credit(origin, local_xcm, hash, weight, weight);
+		Self::deposit_event(Event::Attempted { outcome: outcome.clone() });
+		if let Some(remote_xcm) = remote_xcm {
+			outcome.ensure_complete().map_err(|_| Error::<T>::FeesNotMet)?;
+
+			let (ticket, price) = validate_send::<T::XcmRouter>(dest, remote_xcm.clone())
+				.map_err(Error::<T>::from)?;
+			if origin != Here.into_location() {
```

Comment:
This would need to be fixed for pallet-xcm in general, not just this extrinsic - let's do it in a separate PR - this one is already doing a bit too much if you ask me...

opened https://github.com/paritytech/polkadot-sdk/issues/2082 for it